### PR TITLE
Fix empty lines at the end of the luck/morale description

### LINKED ISSUE #7818: Hero info window, morale modifiers pop-up window is too large
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

With the latest versions morale info window became too big:
image
image

A version a few weeks ago doesn't have such issue:
image

### Save file

None.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by Districh-ru**
Fix #7818
This PR fixes empty lines at the end of the luck/morale modifiers string and also cleans-up 'heroes_indicator' code (luck, morale, experience indicators and description).

morale

### CODE REVIEW

File: src/fheroes2/heroes/heroes_indicator.h [hunk 1]
```diff
@@ -40,47 +43,61 @@ class Heroes;
...
-    void SetPos( const fheroes2::Point & );
+    void SetPos( const fheroes2::Point & pt );
     void SetHero( const Heroes * hero );
 
 protected:
-    const Heroes * hero;
+    const Heroes * _hero;
     fheroes2::Rect area;
-    fheroes2::ImageRestorer back;
+    fheroes2::ImageRestorer back{ fheroes2::Display::instance() };
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_indicator.h:59]: Please assign it within the constructor. This is needed for the future changes when we want to scale a window then we will pass an image to be rendered at.

File: src/fheroes2/heroes/heroes_indicator.h [hunk 2]
```diff
@@ -40,47 +43,61 @@ class Heroes;
 class HeroesIndicator
 {
 public:
-    explicit HeroesIndicator( const Heroes * h = nullptr );
+    explicit HeroesIndicator( const Heroes * hero = nullptr )
+        : _hero( hero )
+    {
+        descriptions.reserve( 256 );
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_indicator.h:49]: Please remove this line and just add `// Do nothing.` comment. Reserving memory for a string makes sense only if the size is known. Here we just make an approximation that might allocate way more memory than needed.

File: src/fheroes2/heroes/heroes_indicator.cpp [hunk 3]
```diff
@@ -129,54 +117,53 @@ void LuckIndicator::QueueEventProcessing( const LuckIndicator & indicator )
...
-    else
+    }
+    else {
         descriptions.append( modificators );
+    }
 
-    descriptions.append( "\n \n" );
-    if ( hero->GetArmy().AllTroopsAreUndead() ) {
+    if ( _hero->GetArmy().AllTroopsAreUndead() ) {
+        descriptions.append( "\n \n" );
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_indicator.cpp:152]: Since we use new Text classes there is not need to add a space in between **\n** symbols.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Districh-ru , I put several comments in this pull request. Could you please take a look at them when you have time?

File: src/fheroes2/heroes/heroes_indicator.h (hunk 1 above)
Comment (Districh-ru) [reply src/fheroes2/heroes/heroes_indicator.h]: Hi, @ihhub, sure)

File: src/fheroes2/heroes/heroes_indicator.h (hunk 2 above)
Comment (Districh-ru) [reply src/fheroes2/heroes/heroes_indicator.h]: OK, I also removed memory reserving for `modificators` string.

File: src/fheroes2/heroes/heroes_indicator.cpp (hunk 3 above)
Comment (Districh-ru) [reply src/fheroes2/heroes/heroes_indicator.cpp]: Done.

**Review by ihhub — APPROVED** [review]

**Review by Branikolog — APPROVED** [review]

**Comment by ihhub** [comment]
@Districh-ru , huge thanks for this fix!

**Pull request merged**